cpu/stm32_common/periph/i2c.c

```diff
@@ -0,0 +1,463 @@
+/*
+ * Copyright (C) 2015 Jan Pohlmann <jan-pohlmann@gmx.de>
+ *               2017 we-sens.com
+ *               2018 Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_stm32l0
+ * @ingroup     drivers_periph_i2c
+ * @{
+ *
+ * @file
+ * @brief       Low-level I2C driver implementation
+ *
+ * @note This implementation only implements the 7-bit addressing polling mode
+ * (for now interrupt mode is not available)
+ *
+ * @author      Peter Kietzmann <peter.kietzmann@haw-hamburg.de>
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @auhtor      Thomas Eichinger <thomas.eichinger@fu-berlin.de>
+ * @author      Jan Pohlmann <jan-pohlmann@gmx.de>
+ * @author      Aurélien Fillau <aurelien.fillau@we-sens.com>
+ * @author      Alexandre Abadie <alexandre.abadie@inria.fr>
+ *
+ * @}
+ */
+
+#include <stdint.h>
+
+#include "cpu.h"
+#include "mutex.h"
+
+#include "cpu_conf_stm32_common.h"
+
+#include "periph/i2c.h"
+#include "periph/gpio.h"
+#include "periph_conf.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define TICK_TIMEOUT    (0xFFFF)
+
+#if defined(CPU_FAM_STM32F0) || defined(CPU_FAM_STM32F3) || \
+    defined(CPU_FAM_STM32F7) || defined(CPU_FAM_STM32L0) || \
+    defined(CPU_FAM_STM32L4)
+
+#define I2C_IRQ_PRIO    (1)
+
+#define CLEAR_FLAG(dev)            dev->ICR |= I2C_ICR_NACKCF   | \
+                                               I2C_ICR_ARLOCF   | \
+                                               I2C_ICR_BERRCF
+
+#define ERROR_FLAG                 (I2C_ISR_NACKF | I2C_ISR_ARLO | I2C_ISR_BERR)
+
+/* static function definitions */
+static inline void _i2c_init(I2C_TypeDef *i2c, uint32_t timing);
+static inline void _start(I2C_TypeDef *dev, uint8_t address, uint8_t length, uint8_t rw_flag);
+static inline int _read(I2C_TypeDef *dev, uint8_t *data, int length);
+static inline int _write(I2C_TypeDef *i2c, const uint8_t *data, int length);
+static inline void _stop(I2C_TypeDef *i2c);
+
+/**
+ * @brief Array holding one pre-initialized mutex for each I2C device
+ */
+static mutex_t locks[I2C_NUMOF];
+
+int i2c_init(i2c_t dev)
+{
+    assert(dev < I2C_NUMOF);
+
+    DEBUG("[i2c] init: initializing device\n");
+    mutex_init(&locks[dev]);
+
+    I2C_TypeDef *i2c = i2c_config[dev].dev;
+
+    periph_clk_en(i2c_config[dev].bus, i2c_config[dev].rcc_mask);
+
+#if defined(CPU_FAM_STM32F0) || defined(CPU_FAM_STM32F3)
+    /* Clear I2CSW bit */
+    RCC->CFGR3 &= ~(i2c_config[dev].rcc_sw_mask);
+    /* Set I2CSW bits according */
+    RCC->CFGR3 |= i2c_config[dev].rcc_sw_mask;
+#endif
+
+    NVIC_SetPriority(i2c_config[dev].irqn, I2C_IRQ_PRIO);
+    NVIC_EnableIRQ(i2c_config[dev].irqn);
+
+    DEBUG("[i2c] init: configuring pins\n");
+    /* configure pins */
+    gpio_init(i2c_config[dev].scl_pin, GPIO_OD_PU);
```

Comment:
Not sure if `GPIO_OD_PU` is the best here, same for sda. Why not using `GPIO_OUT`?